assets/js/src/tracker/index.js

```diff
@@ -0,0 +1,81 @@
+import { config } from '../config';
+import * as formatters from './data-formatting';
+
+let instance;
+
+/**
+ * A tracking utility for initializing a GA4 and tracking accepted events.
+ *
+ * @class
+ */
+class Tracker {
+	/**
+	 * Constructs a new instance of the Tracker class.
+	 *
+	 * @throws {Error} If an instance of the Tracker already exists.
+	 */
+	constructor() {
+		if ( instance ) {
+			throw new Error( 'Cannot instantiate more than one Tracker' );
+		}
+		instance = this;
+		instance.init();
+	}
+
+	/**
+	 * Initializes the tracker and dataLayer if not already done.
+	 */
+	init() {
+		if ( window[ config.tracker_var ] ) {
+			// Tracker already initialized. Do nothing.
+			return;
+		}
+
+		window.dataLayer = window.dataLayer || [];
+
+		function gtag() {
+			window.dataLayer.push( arguments );
+		}
+
+		window[ config.tracker_var ] = gtag;
+
+		gtag( 'js', new Date() );
+		gtag( 'set', `developer_id.${ config.developer_id }`, true );
+		gtag( 'config', config.gtag_id, {
+			allow_google_signals: config.allow_google_signals,
+			link_attribution: config.link_attribution,
+			anonymize_ip: config.anonymize_ip,
+			logged_in: config.logged_in,
+			linker: config.linker,
+			custom_map: config.custom_map,
+		} );
+	}
+
+	/**
+	 * Creates and returns an event handler object for a specified event name.
+	 *
+	 * @param {string} name The name of the event.
+	 * @return {{handler: function(*): void}} An object with a `handler` method for processing and tracking the event.
+	 * @throws {Error} If the event name is not supported.
+	 */
+	event( name ) {
+		/* eslint import/namespace: [ 'error', { allowComputed: true } ] */
+		if ( ! formatters[ name ] ) {
+			throw new Error( `Event ${ name } is not supported.` );
+		}
+
+		return {
+			handler: ( data ) => {
+				if ( config.events.includes( name ) ) {
+					window[ config.tracker_var ](
+						'event',
+						name,
+						formatters[ name ]( data )
```

Comment:
We're checking the presence of `formatters[ name ]` function at the time of `event()` call, but we use it much later, when technically, the function may no longer exist (probably unexpected, but technically possible).
In my opinion, we could make it a bit safer: we could cache the function once it is checked. So, the handler will call the function available at the `event` execution time:
```suggestion
		/* eslint import/namespace: [ 'error', { allowComputed: true } ] */
		const formatter = formatters[ name ];
		if ( typeof formatter !== 'function' ) {
			throw new Error( `Event ${ name } is not supported.` );
		}

		return {
			handler: ( data ) => {
				if ( config.events.includes( name ) ) {
					window[ config.tracker_var ](
						'event',
						name,
						formatter( data )
```

Reply:
Updated in 7fcab41